tests/backend/checkmach/test_signatures_functors.ml

```diff
@@ -0,0 +1,19 @@
+(* Tests showing how [zero_alloc] information from a functor parameter's
+   signature can be used in checking the functor's body.
+
+   For more comprehensive tests of using zero_alloc information from signatures,
+   see [test_signatures_separate_{a,b}.ml]. *)
+
+(* Most basic use case. *)
+module type S_basic = sig
+  val[@zero_alloc] f : int -> int
+end
+
+module F_basic (X : S_basic) = struct
+  let[@zero_alloc] g x = X.f x
+end
+
+(* A non-strict assumption won't help you with a strict check. *)
+module F_strict_bad (X : S_basic) = struct
+  let[@zero_alloc strict] g x = X.f x
+end
```

Comment:
> At minimum shouldn't this produce a warning for a misplaced zero_alloc attribute? (Even better if we could produce a custom error message that said something like "try adding an argument to f".)

This example does seem to give warning 53, as expected.  There are tests for putting [@zero_alloc] on lets that aren't function definitions in `ocaml/testsuite/tests/warnings/w53.ml`.

Separately, warning 53 is not issued when using the toplevel, for any attribute.  Probably we should change that - I can look tomorrow.  (But I think it's an orthogonal issue)

> Even even better if the output f was tracked as [@@zero_alloc], but I figure this might be hard. (But maybe it's not?)

Right now, we have a simple rule for what terms you can put zero_alloc on: function definitions and function applications.  The former case follows exactly the same rules as other function attributes (like `[@inline]`).  I agree we could probably support keeping the information in the case of some aliases, but I think I prefer to have a simple predictable rule about where the attribute is allowed instead, in keeping with the behavior of those other function attributes.  Here, if you want the `f` in Make's output to be zero alloc, you can eta expand it.

Reply:
Yes, I was confused by the top-level thing. Thanks.

I don't think extending the rule to also allow `let[@zero_alloc] ident1 = ident2` threatens its simplicity. Do you think otherwise? 

I'm willing to accept that we don't want to do this at the moment because it would take more time/resources. But I do think it's a natural enough thing to want, and not dramatically difficult to implement.

Reply:
Allowing it would make adding annotations to existing code simpler, it was requrested by users (but not high on the wishlist).

Reply:
I think if we want this, we should do it as a later PR.

Reply:
That seems good to me. Chris, feel free to resolve this comment even if you’re unconvinced we should ever do this.